packages/@lwc/engine-core/src/framework/vm.ts

```diff
@@ -710,3 +711,11 @@ export function forceRehydration(vm: VM) {
         scheduleRehydration(vm);
     }
 }
+
+export function hasShadow(vm: VM): boolean {
+    if (!features.ENABLE_LIGHT_DOM_COMPONENTS) {
+        return true;
+    }
+    const { shadow } = vm.def.ctor;
```

Comment:
I don't think the value should be looked up from the `ctor` directly. The `shadow` property should be looked up in `createComponentDef` and store in the `ComponentDef`. This avoids issues where the field is updated at runtime.